app/main.py

```diff
@@ -1,34 +1,157 @@
+            self.decks = [
+                Deck(start[0], i)
+                for i in range(start[1], end[1] + 1)
+            ]
+        if start[1] == end[1]:
+            self.decks = [
+                Deck(i, start[1])
+                for i in range(start[0], end[0] + 1)
+            ]
+
+    def __repr__(self) -> str:
+        return f"ship: {self.decks}, is_drowned={self.is_drowned}"
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def get_deck(self, row: int, column: int) -> None | Deck:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def fire(self, row: int, column: int) -> str:
+        deck = self.get_deck(row, column)
+        deck.is_alive = False
+        self.is_drowned = all([
+            not deck.is_alive
+            for deck in self.decks
+        ])
+        return "Sunk!" if self.is_drowned else "Hit!"
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list[tuple]) -> None:
+        self.field = {
+            ship: Ship(ship[0], ship[1])
+            for ship in ships
+        }
+        self._validate_field()
+
+    def __str__(self) -> str:
+        return f"{self.field}"
+
+    def _validate_field(self) -> None:
+        if len(self.field) != 10:
+            raise ValidationInputDataError(
+                "The total number of ships should be 10"
+            )
+
+        ships_count = {1: 0, 2: 0, 3: 0, 4: 0}
+        for ship_start, ship_end in self.field.keys():
+            ship_len = 1 + max(
+                abs(ship_start[0] - ship_end[0]),
+                abs(ship_start[1] - ship_end[1])
+            )
+            if ship_len > 4:
+                raise ValidationInputDataError(
+                    "The maximum length of the ship should not exceed 4 decks"
+                )
+            ships_count[ship_len] += 1
+
+        if ships_count != {1: 4, 2: 3, 3: 2, 4: 1}:
+            raise ValidationInputDataError(
+                """There should be 4 single-deck ships
+                There should be 3 double-deck ships
+                There should be 2 three-deck ships
+                There should be 1 four-deck ship"""
+            )
+
+        ship_deck_all = []
+        for ship in self.field.keys():
```

Comment:
when iterating over a dictionary, iterating directly over the dictionary itself (like for ship in self.field) is the same as iterating over the dictionary's keys (like for ship in self.field.keys()). So using keys() is redundant